### DISCUSSION TITLE
Public/static directory lost in multistage docker build. What build assets to copy?

### DISCUSSION THREAD

**Discussion by malimccalla** [vercel/next.js#13304] [Help] [answered]
Using a [simple next js set up](https://github.com/malimccalla/simple-docker-next) (same as [custom-server-typescript](https://github.com/zeit/next.js/tree/canary/examples/custom-server-typescript) but with an added `public` directory) I am trying to build and run it within a docker container. I can get everything working with this Dockerfile

```dockerfile
FROM node:latest
WORKDIR /app

COPY ./package.json ./package.json
COPY ./yarn.lock ./yarn.lock

RUN yarn install

COPY . .

RUN yarn build

CMD ["yarn", "start"]
```

However my app requires a multistage build so I am trying to copy over the build assets with this Dockerfile

```dockerfile
## BUILD PHASE ##
FROM node:latest as builder
WORKDIR /app

COPY ./package.json ./package.json
COPY ./yarn.lock ./yarn.lock

RUN yarn install

COPY . .

RUN yarn run build

## RUN PHASE ##
FROM node:lts-alpine
WORKDIR /app

COPY --from=builder /app/package.json .
COPY --from=builder /app/yarn.lock .

RUN yarn install --production

COPY --from=builder /app/.next ./.next
COPY --from=builder /app/dist ./dist
# Do I need to copy something else from the build output here???

ENV NODE_ENV production
EXPOSE 3000

CMD ["yarn", "start"]
```

This appears to work just fine at first but files within my `public` directory are no longer being served when ran. I'm wondering if I need to copy over some additional folders besides the `dist` and `.next` to get my static/public assets working.

**Comment by malimccalla** [comment, marked as answer]
Solved my issue. I needed to copy over the `public` directory as well, i just assumed it would be built with the project. For anyone that may find it helpful, here is my final working Dockerfile (I have a yarn workspace)

```dockerfile
## BUILD PHASE ##
FROM node:latest as builder
WORKDIR /app

COPY ./lerna.json .
COPY ./package.json .
COPY ./yarn.lock .
COPY ./tsconfig.json .
COPY ./packages/www/package.json ./packages/www/package.json
COPY ./packages/shared/package.json ./packages/shared/package.json

# Will use cache unless any of the above files have changed
RUN yarn install

# Copy the source code to be built
COPY ./packages/www/ ./packages/www/
COPY ./packages/shared/ ./packages/shared/

RUN yarn run build:www

## RUN PHASE ##
FROM node:lts-alpine
WORKDIR /app

COPY --from=builder /app/package.json .
COPY --from=builder /app/packages/www/package.json ./packages/www/package.json
COPY --from=builder /app/packages/shared/package.json ./packages/shared/package.json

RUN yarn install --production

COPY --from=builder /app/packages/shared/dist /app/packages/shared/dist

COPY --from=builder /app/packages/www/dist /app/packages/www/dist
COPY --from=builder /app/packages/www/public /app/packages/www/public
COPY --from=builder /app/packages/www/.next /app/packages/www/.next

WORKDIR /app/packages/www

ENV NODE_ENV production
EXPOSE 3000

CMD ["node", "dist/server.js"]
```

**Comment by cjimmy** [comment]
A note for those seeing this in 2025. When you configure `output: "standalone"` in `next.config.js` (as it does in the[ example containerized app](https://github.com/nextjs/deploy-google-cloud-run)), `public` is intentionally not copied over. See [docs](https://nextjs.org/docs/pages/api-reference/config/next-config-js/output#automatically-copying-traced-files:~:text=This%20minimal%20server,serve%20these%20automatically).
